### PR TITLE
allow up to 4GB per bitmap index

### PR DESCRIPTION
**Pull request by richardstartin**
This doubles the permissible size of a bitmap index in a backward compatible way by using the extra bit which is currently always clear because writing fails when the index size exceeds 2GB. An index size in excess of 2GB isn't recommended, but this is more tolerant than throwing an exception, and is a backward compatible change.

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitmapInvertedIndexReader.java [hunk 1]
```diff
@@ -40,22 +40,25 @@ public class BitmapInvertedIndexReader implements InvertedIndexReader<ImmutableR
...
 
   @SuppressWarnings("unchecked")
   @Override
   public ImmutableRoaringBitmap getDocIds(int dictId) {
-    int offset = _offsetBuffer.getInt(dictId * Integer.BYTES);
-    int length = _offsetBuffer.getInt((dictId + 1) * Integer.BYTES) - offset;
-    return new ImmutableRoaringBitmap(_bitmapBuffer.toDirectByteBuffer(offset - _firstOffset, length));
+    long offset = getOffset(dictId);
+    long length = getOffset(dictId + 1) - offset;
+    return new ImmutableRoaringBitmap(_bitmapBuffer.toDirectByteBuffer(offset - _firstOffset, (int) length));
```
Comment (richardstartin) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitmapInvertedIndexReader.java:57]: Note that the maximum size of a `RoaringBitmap` is 2^16 * (8KB + 2B) ~ 512MB so the cast to int never overflows.

**Comment by richardstartin** [comment]
This isn’t truly backward compatible - all servers need to be upgraded before the first segment with a bitmap index larger than 2GB is completed.

**Review by Jackie-Jiang — APPROVED** [review]
Do we check overflow currently? I feel the old code won't fail on the writer side, but will fail on the reader side when the size is over 2GB?

**Comment by amrishlal** [comment]
> Do we check overflow currently? I feel the old code won't fail on the writer side, but will fail on the reader side when the size is over 2GB?

I don't think overflow is being checked. This is the stack trace I saw while attempting to generate a very large index (2GB+). This was more of a test rather than a production usecase, so we don't really need the limit to be increased, but if it can be done easily then no harm. Although, it would be good to check the overflow and add an error message.

```Error: java.lang.RuntimeException: java.lang.IllegalArgumentException: Negative position at org.apache.pinot.hadoop.job.mappers.SegmentCreationMapper.map(SegmentCreationMapper.java:310) at org.apache.pinot.hadoop.job.mappers.SegmentCreationMapper.map(SegmentCreationMapper.java:66) at org.apache.hadoop.mapreduce.Mapper.run(Mapper.java:146) at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:793) at org.apache.hadoop.mapred.MapTask.run(MapTask.java:341) at org.apache.hadoop.mapred.YarnChild$2.run(YarnChild.java:177) at java.security.AccessController.doPrivileged(Native Method) at javax.security.auth.Subject.doAs(Subject.java:422) at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1893) at org.apache.hadoop.mapred.YarnChild.main(YarnChild.java:171) Caused by: java.lang.IllegalArgumentException: Negative position at sun.nio.ch.FileChannelImpl.map(FileChannelImpl.java:863) at org.apache.pinot.segment.local.segment.creator.impl.inv.BitmapInvertedIndexWriter.mapBitmapBuffer(BitmapInvertedIndexWriter.java:102) at org.apache.pinot.segment.local.segment.creator.impl.inv.BitmapInvertedIndexWriter.resizeIfNecessary(BitmapInvertedIndexWriter.java:95) at org.apache.pinot.segment.local.segment.creator.impl.inv.BitmapInvertedIndexWriter.add(BitmapInvertedIndexWriter.java:73) at org.apache.pinot.segment.local.segment.creator.impl.inv.json.OnHeapJsonIndexCreator.seal(OnHeapJsonIndexCreator.java:57) at org.apache.pinot.segment.local.segment.creator.impl.SegmentColumnarIndexCreator.seal(SegmentColumnarIndexCreator.java:560) at org.apache.pinot.segment.local.segment.creator.impl.SegmentIndexCreationDriverImpl.handlePostCreation(SegmentIndexCreationDriverImpl.java:266) at org.apache.pinot.segment.local.segment.creator.impl.SegmentIndexCreationDriverImpl.build(SegmentIndexCreationDriverImpl.java:238) at org.apache.pinot.hadoop.job.mappers.SegmentCreationMapper.map(SegmentCreationMapper.java:277) ... 9 more Suppressed: java.lang.IllegalArgumentException: Negative size at sun.nio.ch.FileChannelImpl.truncate(FileChannelImpl.java:324) at org.apache.pinot.segment.local.segment.creator.impl.inv.BitmapInvertedIndexWriter.close(BitmapInvertedIndexWriter.java:118) at org.apache.pinot.segment.local.segment.creator.impl.inv.json.OnHeapJsonIndexCreator.seal(OnHeapJsonIndexCreator.java:50) ... 13 more```

**Comment by richardstartin** [comment]
> Do we check overflow currently? I feel the old code won't fail on the writer side, but will fail on the reader side when the size is over 2GB?

Currently the code fails on the writer side, but would also fail on the reader side. The right way to evolve this is to make these changes, but continue failing beyond 2GB on the writer side, then increase to 4GB the version after.

**Comment by amrishlal** [comment]
> This isn’t truly backward compatible - all servers need to be upgraded before the first segment with a bitmap index larger than 2GB is completed.

I am wondering how this would work? We deploy new versions of servers on a cluster and then once all servers are on the new version, we deploy a config update to switch realtime and offline segment generation to allow for reading/writing 4GB index files?

**Review by amrishlal — APPROVED** [review]
It will be good to add some comments for future reference, but otherwise looks good to me. The compatibility issue doesn't appear to be concern (at least on the clusters I work with) since all the index files are less than 2GB.

**Comment by siddharthteotia** [comment]
Shall we add a test where we generate 2 bitmap inv indexes > 2GB and < 2GB respectively and the reader code changed in this PR can read both indexes correctly ?

**Comment by richardstartin** [comment]
> Shall we add a test where we generate 2 bitmap inv indexes > 2GB and < 2GB respectively and the reader code changed in this PR can read both indexes correctly ?

< 2GB is already tested. Tests could be added for > 2GB, with the downside of needing to create large files.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitmapInvertedIndexReader.java (hunk 1 above)
Comment (jackjlli) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitmapInvertedIndexReader.java:57]: It'd be good to add this comments of analysis to the code above as well?

**Comment by richardstartin** [comment]
@jackjlli this is not backward incompatible

**Comment by jackjlli** [comment]
> This isn’t truly backward compatible - all servers need to be upgraded before the first segment with a bitmap index larger than 2GB is completed.

I added the label as you mentioned the following sentence:
```
This isn’t truly backward compatible - all servers need to be upgraded before the first segment with a bitmap index larger than 2GB is completed.
```
I think the label is used for identifying any action needed if a newer version of Pinot is bumped in the service, so that ppl are aware of those actions.

**Comment by richardstartin** [comment]
Segments produced by the existing code can still be consumed, but the existing code isn’t _forward compatible_ with a limit increase. This change can be made safely in two stages as mentioned above: 1) release these code changes but enforce a 2GB limit, which makes the reader code forward compatible with a limit increase but can’t produce unreadable segments, then 2) relax the 2GB limit in the next release. This just requires upgrading users not to skip versions.

**Pull request merged**